text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
@tari Indeed there aren’t that many articles about that out there. Most of it appears to be a kind of insider knowledge. [Here](http://bysin.net/evidence-based-research/2017/03/04/comparing-the-performance-of-synchronous-versus-asynchronous-network-io/)’s a bit more recent benchmark though, which appears to confirm the cliche of the benefits of asynchronous IO.

Basically in most cases synchronous IO is often better on all metrics for lower concurrencies of less than around 10 threads per CPU core or so (in my experience), and still often maintains a higher data throughput than asynchronous IO even beyond that limit.

The downsides of synchronous IO is the increased burden by the usually (necessarily) very complex kernel scheduler which shows significantly higher CPU and memory usage (e.g. at least 64 instead of 4KB _RSS_ per connection on Linux) on very high levels of concurrency.
IMO most servers can absolutely be written with sync IO and plain threads without any noticeable downsides. I wouldn’t use it for a web facing servers (see Apache’s performance) or e.g. a WebSocket server which is naturally often highly concurrent though. But Apache is still widely spread so there's that…

It's quite a simple reasoning really: A user land scheduler does less and thus needs less. I don't think this will change in the future for all relevant kernels.

BTW Regarding the limited address space: x86_64 CPUs have AFAIK an addressable range of  128TB, which results in just 16k possible concurrent threads if each of them had the common 8MB stack space. So it's not like that's a thing of the past right?

<hr>

The main benefit of stackless coroutines is the fact they don't have any overhead compared to just synchronous code. They're also easier to implement and allow you to write things like generators/iterators more comfortably (i.e. using the well known yield syntax). It's downsides compared to stackful coroutines still hold true, but considering the use cases I just mentioned, stackless ones IMO do have a very much justified place in most languages.